src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,712 @@
+            }
+            return result;
+        }
+    }
+
+    public static class KeyFactoryImpl extends KeyFactorySpi {
+        @Override
+        protected PublicKey engineGeneratePublic(KeySpec keySpec) throws InvalidKeySpecException {
+            if (keySpec instanceof X509EncodedKeySpec x) {
+                try {
+                    var val = new DerValue(new ByteArrayInputStream(x.getEncoded()));
+                    val.data.getDerValue();
+                    return new HSSPublicKey(new DerValue(val.data.getBitString()).getOctetString());
+                } catch (IOException | InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            } else if (keySpec instanceof RawKeySpec x) {
+                try {
+                    return new HSSPublicKey(x.getKeyArr());
+                } catch (InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            }
+            throw new InvalidKeySpecException("Unrecognized KeySpec");
+        }
+
+        @Override
+        protected PrivateKey engineGeneratePrivate(KeySpec keySpec) throws InvalidKeySpecException {
+            throw new InvalidKeySpecException("Private key generation is not supported");
+        }
+
+        @Override
+        protected <T extends KeySpec> T engineGetKeySpec(Key key, Class<T> keySpec) throws InvalidKeySpecException {
+            if (key == null) {
+                throw new InvalidKeySpecException("key should not be null");
+            }
+            if (key.getFormat().equals("X.509") && key.getAlgorithm().equalsIgnoreCase("HSS/LMS")) {
+                if (keySpec.isAssignableFrom(X509EncodedKeySpec.class)) {
+                    return keySpec.cast(new X509EncodedKeySpec(key.getEncoded()));
+                }
+                throw new InvalidKeySpecException("keySpec is not an X509 one");
+            }
+            throw new InvalidKeySpecException("Wrong key format or key algorithm");
+        }
+
+        @Override
+        protected Key engineTranslateKey(Key key) throws InvalidKeyException {
+            if (key == null) {
+                throw  new InvalidKeyException("key cannot be null");
+            }
+            PublicKey pKey;
+            try {
+                // Check if key originates from this factory
+                if (key instanceof HSSPublicKey) {
+                    return key;
+                }
+                // Convert key to spec
+                X509EncodedKeySpec x509EncodedKeySpec
+                        = engineGetKeySpec(key, X509EncodedKeySpec.class);
+                // Create key from spec, and return it
+                pKey = engineGeneratePublic(x509EncodedKeySpec);
+            } catch (InvalidKeySpecException e) {
+                throw new InvalidKeyException(e);
+            }
+            return pKey;
+        }
+    }
+
+    static class HSSPublicKey extends X509Key implements Length {
```

Comment:
Is there a significant value in extending `X509Key`? I think that is primarily to help in the parsing by calling `super()`, but you are not using that. You could override `toString()`, `getFormat()`, `getEncoded()`, etc and probably get the same functionality. @wangweij any thoughts on this?